### ISSUE TITLE
[Feature request] allow sorting stacked series in the same order as legend

### ISSUE THREAD

**Issue by lukaszgryglicki** [grafana/grafana#9538]
I have a stacked chart that uses multi-value template variable to select which series to show:
https://devstats.k8s.io/dashboard/db/companies-stats?orgId=1
I can sort legend by min, max, avg etc.
But stacked series are displayed in alphabetical order, so, for example, Google which is the biggest part of all stacked series is not 1st, Red Hat is not 2nd etc. 

I would like to have an option to sort stacked series in the same order as current order of legend
 
Screenshot:
![screen shot 2017-10-16 at 08 42 45](https://user-images.githubusercontent.com/2469783/31598440-0225e702-b24e-11e7-8124-3a6bb1f80dee.png)

Would be really great to have some checkbox option saying "sort stacked series the same way as legend" or something similar.

**Comment by bergquist** [comment]
This is currently not possible at a Grafana level but some datasources allows you to sort the result. ex http://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.sortByTotal

But it would make sense to add this to Grafana. We will take a look at it.

**Comment by lukaszgryglicki** [comment]
Thanks!

**Comment by dankohn** [comment]
@bergquist Should we open an issue with InfluxDB? We also have a lot of folks interested in the devstats project there. But what's the fastest way to get this implemented? What's the best way?

**Comment by lukaszgryglicki** [comment]
I think we need that on Grafana, because even if template variable values are sorted the way we want, Grafana will still show series in alphabetical order.
Even if InfluxDB would support that, then we would have to use different query each time we change legend sort order.
I think adding this in Grafana would be the best solution and I can imagine that such a feature would be very useful for everybody.

**Comment by bergquist** [comment]
@dankohn I might be mistaken that Influxdb does not support ordering the result based on values. So double checking with them sounds like a good idea. 

I'm working on this feature at the moment and it should be ready for review tomorrow/tonight (depending on when my kid falls a sleep :)). Once its merged to master we provide a build for every successful commit.

**Comment by lukaszgryglicki** [comment]
Great!! Thanks a lot!
How can I get this new feature to allow it on our test server?

**Comment by bergquist** [comment]
@lukaszgryglicki We build every successful commit to master and publish it as..
* packages at https://grafana.com/grafana/download (choose the night version in the upper right) 
* docker image on https://hub.docker.com/r/grafana/grafana/ with the master tag. 

So wait ~20min and download it from grafana.com or purge your local docker image and pull it again :)

**Comment by lukaszgryglicki** [comment]
Thx.
I'll try it on a test server (https://cncftest.io) tomorrow.
Will play a little with it, and if all is fine, I'll deploy on prod server (https://devstats.k8s.io).
Once again thanks for a great job! :D

**Comment by bergquist** [comment]
@lukaszgryglicki Its always a good idea to backup the database before you upgrade Grafana. just in case something breaks. Otherwise its just download, install and start the new version :)

Right now you might experience some problems with plugins since its a night build. Otherwise it should work as expected.

CNCF not running docker?! ;)

**Comment by lukaszgryglicki** [comment]
No problem I'll backup `grafana.db`.
We were using Docker initially, but to make all things faster we decided to put all tools locally on the server and make them communicate locally (devstats, postgres, influxdata, apache proxy, grafana...)
;-)

**Comment by bergquist** [comment]
btw! Grafana now also supports Postgres nativly. So if you want to write queries with joins you can query postgres directly. But Influxdb is still gonna be faster for simple queries.

**Comment by dankohn** [comment]
We are aiming to move containerize the software but we ran into some bugs
and didn't want to hold off on deployment. We aim to circle back and get
things containerized.

--
Dan Kohn <dan@linuxfoundation.org>
Executive Director, Cloud Native Computing Foundation https://www.cncf.io
+1-415-233-1000 https://www.dankohn.com

**Comment by lukaszgryglicki** [comment]
Postgres queries are complex and not suitable to use them directly, I think the best way is to call them once per hour and to save their results to simplest possible Influx series (GitHub archives are published once per hour).
For example, that query would be slow when called directly from Grafana once for all data periods:
https://github.com/cncf/devstats/blob/master/metrics/prs_blocked.sql
https://devstats.k8s.io/dashboard/db/blocked-prs?orgId=1

**Comment by lukaszgryglicki** [comment]
I've installed new Grafana nightly build.
Feature works. Thanks a lot :)

I'm thinking that maybe the order should be reversed? (I'm not sure)
For example when sorting by avg desc on the legend - Kubernetes which has the highest avg is stacked last (its value is on the top).
https://cncftest.io/dashboard/db/approvers?orgId=1

Or maybe there should be some UI to control this.... I remember that at some point (during working on this feature) there was a UI config for this, but I cannot see this in final nightly build (I've also taken a look at the changed files in a merge commit). 
Would be great to have UI saying: sort the same way as the legend, reverse and none.

This is how dashboards look now:
https://cncftest.io/dashboard/db/bot-commands?orgId=1
https://cncftest.io/dashboard/db/companies-stats?orgId=1
https://cncftest.io/dashboard/db/companies-velocity?orgId=1
https://cncftest.io/dashboard/db/issues-repository-group?orgId=1
https://cncftest.io/dashboard/db/prs-approval?orgId=1
https://cncftest.io/dashboard/db/prs-approval-stacked?orgId=1
https://cncftest.io/dashboard/db/prs-merged?orgId=1
https://cncftest.io/dashboard/db/prs-merged-repository-groups?orgId=1
https://cncftest.io/dashboard/db/sig-issues?orgId=1
https://cncftest.io/dashboard/db/sig-mentions?orgId=1
https://cncftest.io/dashboard/db/sig-mentions-categories?orgId=1
https://cncftest.io/dashboard/db/sig-mentions-using-labels?orgId=1

I'll do some more tweaks and tests.

**Comment by lukaszgryglicki** [comment]
Installed on https://devstats.k8s.io after some testing on https://cncftest.io.

Once again thanks for a new feature!

**Comment by bergquist** [comment]
We removed the option since it makes sense to always enable sorting of series when having stack series and sorted legend. Our goal is to introduce as few options as possible. Especially when we can find conventions that make sense instead. (convention over configuration :heart: )

I think your right about inverting the sort order. Fix incomming

**Comment by dankohn** [comment]
Inverting the sort order by default would be good, but even more useful is that when the user inverts the sort order (by clicking on `avg`) the URL should change.

Stateful URLs are a huge strength of Grafana, so it's surprising and disappointing when they don't work.

**Comment by bergquist** [comment]
Agree. Definitely something we want to support. Not sure when we will be able to include it in our road map.

**Comment by bergquist** [comment]
ref https://github.com/grafana/grafana/issues/9617

**Comment by lukaszgryglicki** [comment]
Is this reversed sort build ready? Or should I wait for #9617 ?

**Comment by bergquist** [comment]
Its included in our night build (our master is stable). https://grafana.com/grafana/download/4.7.0-9930pre1

I would not wait for #9617 we don't know if or when we will start working on that yet.

**Comment by lukaszgryglicki** [comment]
OK, thanks, I'll let You know when I update Grafana, now I need to finish some started stuff.
Thanks!

**Comment by lukaszgryglicki** [comment]
New build is broken at least for me, I have no more time for it but it fails very often with those errors:

On those pages (refresh usually help), I, of course, have annotations defined OK:
https://cncftest.io/dashboard/db/issues-age?orgId=1
https://cncftest.io/dashboard/db/dashboards?refresh=15m&orgId=1
I had to reimport all dashboards manually now :(

BTW: is the a way to batch import JSONs (like 35 dashboards JSONs) from commandline on the server?
![screen shot 2017-10-27 at 16 54 59](https://user-images.githubusercontent.com/2469783/32113075-773f167c-bb3f-11e7-8156-08d1682a5d6c.png)
![screen shot 2017-10-27 at 17 46 04](https://user-images.githubusercontent.com/2469783/32113076-775bb9d0-bb3f-11e7-86c8-a6730c535312.png)
![screen shot 2017-10-27 at 17 46 15](https://user-images.githubusercontent.com/2469783/32113077-77762eaa-bb3f-11e7-8f08-7caaef2d2e25.png)
![screen shot 2017-10-27 at 17 46 38](https://user-images.githubusercontent.com/2469783/32113078-77939724-bb3f-11e7-8d8e-7b06e1ee655c.png)
![screen shot 2017-10-27 at 17 47 30](https://user-images.githubusercontent.com/2469783/32113079-77add652-bb3f-11e7-8fb7-4bf23f46a0d2.png)

**Comment by lukaszgryglicki** [comment]
Managed to get rid of all but "undefined is not an object"
It always happens when I first click "Dashboards \/" menu (right to the main menu with kubernetes icon) and choose "Home" and then any other dashboard from the same menu. It doesn't happen when I don't choose "Home" on the way.
Error says something about `this.$timeout` object being incorrect and the Home dashboard is the only one that uses auto-refresh - this suggests some timer related issues (like missing npm/jquery module etc.)....

**Comment by torkelo** [comment]
Fixed the issue with `undefined is not an object evaluating this.$timeout` , caused by a resent js > typescript refactor in master (did not affect the recent 4.6 release). Fixed in latest master/nightly build

**Comment by lukaszgryglicki** [comment]
Great, will retry on Monday and report here.
Thanks.

**Comment by lukaszgryglicki** [comment]
Forgot to mention.... all is working good on a newer build.

**Comment by slseed1969** [comment]
Has this been released? I'm running 5.2.2 and I don't see it in the changelog. Also, I'm not seeing this enabled because my stacked graphs are still sorted alphabetically when my legend is sorted by value.

**Comment by marefr** [comment]
@slseed1969 this was included in v5.0.0-beta1 and later v5.0.0 stable. We're not always adding everything to the changelog. But as you can see on this issue it's in milestone 5.0.0-beta1.

**Comment by slseed1969** [comment]
Thanks. I'm asking because my stacked graphs are not being sorted. I wonder if this is a regression.

**Comment by marefr** [comment]
@slseed1969 works as expected [here](https://k8s.devstats.cncf.io/d/8/company-statistics-by-repository-group?orgId=1). Please make sure to order legend. If you still have problem I would suggest open a new issue describing your problem is as much detail as possible.

**Comment by slseed1969** [comment]
OK. I was able to get my graphs to sort, but only after I added the value columns and changed my legend to table, then clicked on the header to sort. In my graph, my legend was originally not a table and I didn't display the values. I was expecting the stack sort order to match the sort order in the Display Hover tooltip. This does not appear to be the case. Should I open an issue to request this?

**Comment by ernstp** [comment]
I really would like to have option to have the same sort order in the legend as in the stacked graph.
To produce something like this: 
![bild](https://user-images.githubusercontent.com/877338/71099358-6d9ba180-21b3-11ea-9ae0-90ff204c2928.png)

Right now you get the legend upside down.

**Comment by tjanson** [comment]
It seems this no longer works; according to [these forum discussions](https://community.grafana.com/t/sorting-of-stacked-series-in-time-series-panel/69577) it broke somewhere between v. 8.3 and 8.5.

**Comment by AntoineThebaud** [comment]
Can someone reopen this issue? As stated above the problem is back.

**Comment by lukaszgryglicki** [comment]
I'm an author, but I cannot reopen :/
